runners_manager/runner/RunnerManager.py

```diff
@@ -38,40 +39,49 @@ def __init__(self, org: str, config: list,
                 self.create_runner(t)
 
     def update(self, github_runners: list[dict]):
+        # Update status of each runner
         for elem in github_runners:
             runner = self.runners[elem['name']]
             runner.update_status(elem)
 
             if runner.action_id is None:
                 runner.action_id = elem['id']
 
-        for vm_type in self.runner_management:
-            current_online = len(
-                self.filter_runners(vm_type, lambda r: not r.has_run and r.action_id)
-            )
-            offlines = self.filter_runners(vm_type, lambda r: r.has_run)
+        self.log_runners_infos()
 
-            logger.info("type" + str(vm_type))
-            logger.info(f"Currently online: {current_online}")
-            logger.debug('Online runners')
-            logger.debug(','.join([f"{elem.name} {elem.status}"
-                                   for elem in self.filter_runners(vm_type)]))
-            logger.debug('Offline runners')
-            logger.debug(','.join([f"{elem.name} {elem.status}" for elem in offlines]))
+        # runner logic For each type of VM
+        for vm_type in self.runner_management:
+            # Always Respawn Vm
+            offline_runners = self.filter_runners(vm_type, lambda r: r.has_run)
+            for r in offline_runners:
+                self.respawn_replace(r)
 
+            # Create if it's still not enough
             while self.need_new_runner(vm_type):
                 self.create_runner(vm_type)
 
-            if current_online > 0 and len(offlines) > 0:
-                for elem in offlines:
+            # Delete if you have to many
+            # if we can't delete the runner because of Github respawn it and try the next time ?
+            for elem in self.filter_runners(vm_type, lambda r: r.status == 'online'):
+                if datetime.datetime.now() - elem.started_at > datetime.timedelta(hours=2) \
+                        and not self.need_new_runner(vm_type):
```

Comment:
So now if the VM is online for 2 hours you delete it, could you set a global var for the `2` hours, so that it can be tweaked if needed.